--- src/dwm-6.0/dwm.c	2013-03-08 18:43:08.565408608 +0100
+++ b/dwm-6.0/dwm.c	2013-03-08 18:42:28.615409280 +0100
@@ -1775,21 +1775,25 @@
 	if(n == 0)
 		return;
 
-	if(n > m->nmasters[m->curtag])
-		mw = m->nmasters[m->curtag] ? m->ww * m->mfacts[m->curtag] : 0;
-	else
-		mw = m->ww;
-	for(i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
-		if(i < m->nmasters[m->curtag]) {
-			h = (m->wh - my) / (MIN(n, m->nmasters[m->curtag]) - i);
-			resize(c, m->wx, m->wy + my, mw - (2*c->bw), h - (2*c->bw), False);
-			my += HEIGHT(c);
-		}
-		else {
-			h = (m->wh - ty) / (n - i);
-			resize(c, m->wx + mw, m->wy + ty, m->ww - mw - (2*c->bw), h - (2*c->bw), False);
-			ty += HEIGHT(c);
-		}
+	if(n == 1) {
+		monocle(selmon);
+	} else {
+		if(n > m->nmasters[m->curtag])
+			mw = m->nmasters[m->curtag] ? m->ww * m->mfacts[m->curtag] : 0;
+		else
+			mw = m->ww;
+		for(i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
+			if(i < m->nmasters[m->curtag]) {
+				h = (m->wh - my) / (MIN(n, m->nmasters[m->curtag]) - i);
+				resize(c, m->wx, m->wy + my, mw - (2*c->bw), h - (2*c->bw), False);
+				my += HEIGHT(c);
+			}
+			else {
+				h = (m->wh - ty) / (n - i);
+				resize(c, m->wx + mw, m->wy + ty, m->ww - mw - (2*c->bw), h - (2*c->bw), False);
+				ty += HEIGHT(c);
+			}
+	}
 }
 
 void
@@ -2225,22 +2229,25 @@
 	for(n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), n++);
 	if(n == 0)
 		return;
-
-	if(n > m->nmasters[m->curtag])
-		mh = m->nmasters[m->curtag] ? m->wh * m->mfacts[m->curtag] : 0;
-	else
-		mh = m->wh;
-	for(i = mx = tx = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
-		if(i < m->nmasters[m->curtag]) {
-			w = (m->ww - mx) / (MIN(n, m->nmasters[m->curtag]) - i);
-			resize(c, m->wx + mx, m->wy, w - (2*c->bw), mh - (2*c->bw), False);
-			mx += WIDTH(c);
-		}
-		else {
-			w = (m->ww - tx) / (n - i);
-			resize(c, m->wx + tx, m->wy + mh, w - (2*c->bw), m->wh - mh - (2*c->bw), False);
-			tx += WIDTH(c);
-		}
+	if(n == 1) {
+		monocle(selmon);
+	} else {
+		if(n > m->nmasters[m->curtag])
+			mh = m->nmasters[m->curtag] ? m->wh * m->mfacts[m->curtag] : 0;
+		else
+			mh = m->wh;
+		for(i = mx = tx = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
+			if(i < m->nmasters[m->curtag]) {
+				w = (m->ww - mx) / (MIN(n, m->nmasters[m->curtag]) - i);
+				resize(c, m->wx + mx, m->wy, w - (2*c->bw), mh - (2*c->bw), False);
+				mx += WIDTH(c);
+			}
+			else {
+				w = (m->ww - tx) / (n - i);
+				resize(c, m->wx + tx, m->wy + mh, w - (2*c->bw), m->wh - mh - (2*c->bw), False);
+				tx += WIDTH(c);
+			}
+	}
 }
 
 int
