--- dwm.c.orig	2012-05-26 15:51:24.462900696 +0200
+++ src/dwm-6.0/dwm.c	2012-05-26 15:49:41.169569734 +0200
@@ -120,7 +120,8 @@
 	int x, y, w, h;
 	XftColor norm[ColLast];
 	XftColor sel[ColLast];
-	Drawable drawable;
+	XftColor urg[ColLast];
+    Drawable drawable;
 	GC gc;
 	struct {
 		int ascent;
@@ -842,7 +843,7 @@
 	dc.x = 0;
 	for(i = 0; i < LENGTH(tags); i++) {
         dc.w = TEXTW(tags[i].name);
-		col = m->tagset[m->seltags] & 1 << i ? dc.sel : dc.norm;
+        col = urg & 1 << i ? dc.urg : m->tagset[m->seltags] & 1 << i ? dc.sel : dc.norm;
 		drawtext(tags[i].name, col, urg & 1 << i);
 		drawsquare(m == selmon && selmon->sel && selmon->sel->tags & 1 << i,
 		           occ & 1 << i, urg & 1 << i, col);
@@ -1795,7 +1796,10 @@
 	dc.sel[ColBorder] = getcolor(selbordercolor);
 	dc.sel[ColBG] = getcolor(selbgcolor);
 	dc.sel[ColFG] = getcolor(selfgcolor);
-	dc.drawable = XCreatePixmap(dpy, root, DisplayWidth(dpy, screen), bh, DefaultDepth(dpy, screen));
+	dc.urg[ColBorder] = getcolor(urgbordercolor);
+    dc.urg[ColBG] = getcolor(urgbgcolor);
+    dc.urg[ColFG] = getcolor(urgfgcolor);
+    dc.drawable = XCreatePixmap(dpy, root, DisplayWidth(dpy, screen), bh, DefaultDepth(dpy, screen));
 	dc.gc = XCreateGC(dpy, root, 0, NULL);
 	XSetLineAttributes(dpy, dc.gc, 1, LineSolid, CapButt, JoinMiter);
     /* init system tray */
@@ -2334,9 +2338,12 @@
 			wmh->flags &= ~XUrgencyHint;
 			XSetWMHints(dpy, c->win, wmh);
 		}
-		else
+		else {
 			c->isurgent = (wmh->flags & XUrgencyHint) ? True : False;
-		if(wmh->flags & InputHint)
+            if(c->isurgent)
+                XSetWindowBorder(dpy, c->win, dc.urg[ColBorder].pixel);
+        }
+        if(wmh->flags & InputHint)
 			c->neverfocus = !wmh->input;
 		else
 			c->neverfocus = False;
