--- dwm.c	2013-03-19 01:10:27.509901099 +0100
+++ dwm.c	2013-03-19 01:18:53.909892580 +0100
@@ -183,6 +183,7 @@
 static void configurenotify(XEvent *e);
 static void configurerequest(XEvent *e);
 static Monitor *createmon(void);
+static void cycle(const Arg *arg);
 static void destroynotify(XEvent *e);
 static void detach(Client *c);
 static void detachstack(Client *c);
@@ -237,10 +238,12 @@
 static void setlayout(const Arg *arg);
 static void setmfact(const Arg *arg);
 static void setup(void);
+static int shifttag(int dist);
 static void showhide(Client *c);
 static void sigchld(int unused);
 static void spawn(const Arg *arg);
 static void tag(const Arg *arg);
+static void tagcycle(const Arg *arg);
 static void tagmon(const Arg *arg);
 static int textnw(const char *text, unsigned int len);
 static void tile(Monitor *);
@@ -774,6 +777,13 @@
 }
 
 void
+cycle(const Arg *arg) {
+	const Arg a = { .i = shifttag(arg->i) };
+
+	view(&a);
+}
+
+void
 destroynotify(XEvent *e) {
 	Client *c;
 	XDestroyWindowEvent *ev = &e->xdestroywindow;
@@ -1815,6 +1825,17 @@
 	grabkeys();
 }
 
+int
+shifttag(int dist) {
+	int seltags = selmon->tagset[selmon->seltags] & TAGMASK;
+
+	if(dist > 0) /* left circular shift */
+		seltags = (seltags << dist) | (seltags >> (LENGTH(tags) - dist));
+	else /* right circular shift */
+		seltags = (seltags >> (- dist)) | (seltags << (LENGTH(tags) + dist));
+	return seltags;
+}
+
 void
 showhide(Client *c) {
 	if(!c)
@@ -1861,6 +1882,14 @@
 }
 
 void
+tagcycle(const Arg *arg) {
+	const Arg a = { .i = shifttag(arg->i) };
+
+	tag(&a);
+	view(&a);
+}
+
+void
 tagmon(const Arg *arg) {
 	if(!selmon->sel || !mons->next)
 		return;
