--- dwm.c	2013-03-07 20:44:58.959948770 +0100
+++ dwm.c	2013-03-07 20:44:50.559948911 +0100
@@ -189,7 +189,6 @@
 static void drawbar(Monitor *m);
 static void drawbars(void);
 static void drawcoloredtext(char *text);
-static void drawsquare(Bool filled, Bool empty, unsigned long col[ColLast]);
 static void drawtext(const char *text, unsigned long col[ColLast], Bool pad);
 static void enternotify(XEvent *e);
 static void expose(XEvent *e);
@@ -765,9 +764,8 @@
 		if (!(m->tagset[m->seltags] & 1 << i) && !(occ & 1 << i))
 			continue;
 		dc.w = TEXTW(tags[i].name);
-		col = dc.colors[(m->tagset[m->seltags] & 1 << i ? 1:(urg & 1 << i ? 2:0))];
+		col = dc.colors[(m->tagset[m->seltags] & 1 << i) ? 1 : (urg & 1 << i ? 2:(occ & 1 << i ? 3:0)) ];
 		drawtext(tags[i].name, col, True);
-		drawsquare(False, False, urg & 1 << i, col);
 		dc.x += dc.w;
 	}
 	dc.w = blw = TEXTW(m->ltsymbol);
@@ -837,17 +835,6 @@
 }
 
 void
-drawsquare(Bool filled, Bool empty, unsigned long col[ColLast]) {
-	int x;
-	XSetForeground(dpy, dc.gc, col[ColFG]);
-	x = (dc.font.ascent + dc.font.descent + 2) / 4;
-	if(filled)
-		XFillRectangle(dpy, dc.drawable, dc.gc, dc.x+1, dc.y+1, x+1, x+1);
-	else if(empty)
-		XDrawRectangle(dpy, dc.drawable, dc.gc, dc.x+1, dc.y+1, x, x);
-}
-
-void
 drawtext(const char *text, unsigned long col[ColLast], Bool pad) {
 	char buf[256];
 	int i, x, y, h, len, olen;
