--- dwm.c.orig	2013-03-13 01:44:46.546471052 +0100
+++ b/dwm-6.0/dwm.c	2013-03-13 01:47:42.903134752 +0100
@@ -454,13 +454,13 @@
 		for(i = mx = tx = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
 			if(i < m->nmasters[m->curtag]) {
 				w = (m->ww - mx) / (MIN(n, m->nmasters[m->curtag]) - i);
-				resize(c, m->wx + mx, m->wy, w - (2*c->bw), mh - (2*c->bw), False);
-				mx += WIDTH(c);
+				resize(c, m->wx + mx - c->bw, m->wy, w, mh - c->bw, False);
+				mx += WIDTH(c) - c->bw;
 			}
 			else {
 				w = (m->ww - tx) / (n - i);
-				resize(c, m->wx + tx, m->wy + mh, w - (2*c->bw), m->wh - mh - (2*c->bw), False);
-				tx += WIDTH(c);
+				resize(c, m->wx + tx - c->bw, m->wy + mh, w, m->wh - mh - c->bw, False);
+				tx += WIDTH(c) - c->bw;
 			}
 	}
 }
@@ -524,7 +524,7 @@
 	/* buddy list */
 	ww = (bl ? m->mfacts[m->curtag] * m->ww : m->ww);
 	if(bl) {
-		resize(bl, m->wx + ww, m->wy, m->ww - ww - 2 * bl->bw, m->wh - 2 * bl->bw, False);
+		resize(bl, m->wx + ww, m->wy, m->ww - ww, m->wh - bl->bw, False);
 		if(--n == 0)
 			return;
 	}
@@ -554,7 +554,7 @@
 		if(n == 1 && !bl)
 			resize(c, cx - c->bw, cy, cw, ch, False);
 		else
-			resize(c, cx, cy, cw - 2 * c->bw, ch - 2 * c->bw, False);
+			resize(c, cx - c->bw, cy, cw, ch - c->bw, False);
 		rn++;
 		if(rn >= rows) {
 			rn = 0;
@@ -990,6 +990,7 @@
 		attachstack(c);
 		grabbuttons(c, True);
 		XSetWindowBorder(dpy, c->win, dc.colors[1][ColBorder]);
+		XRaiseWindow(dpy, c->win);
 		setfocus(c);
 	}
 	else
@@ -1853,13 +1854,13 @@
 		for(i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
 			if(i < m->nmasters[m->curtag]) {
 				h = (m->wh - my) / (MIN(n, m->nmasters[m->curtag]) - i);
-				resize(c, m->wx, m->wy + my, mw - (2*c->bw), h - (2*c->bw), False);
-				my += HEIGHT(c);
+				resize(c, m->wx - c->bw, m->wy + my, mw - c->bw, h - c->bw, False);
+				my += HEIGHT(c) - c->bw;
 			}
 			else {
 				h = (m->wh - ty) / (n - i);
-				resize(c, m->wx + mw, m->wy + ty, m->ww - mw - (2*c->bw), h - (2*c->bw), False);
-				ty += HEIGHT(c);
+				resize(c, m->wx + mw - c->bw, m->wy + ty, m->ww - mw, h - c->bw, False);
+				ty += HEIGHT(c) - c->bw;
 			}
 	}
 }
