--- dwm.c.orig	2013-03-21 12:34:03.138493187 +0100
+++ b/dwm-6.0/dwm.c	2013-03-21 12:34:50.975159031 +0100
@@ -222,6 +222,7 @@
 static void pushdown(const Arg *arg);
 static void pushup(const Arg *arg);
 static void quit(const Arg *arg);
+static void raisefloating(void);
 static Monitor *recttomon(int x, int y, int w, int h);
 static void resize(Client *c, int x, int y, int w, int h, Bool interact);
 static void resizeclient(Client *c, int x, int y, int w, int h);
@@ -457,13 +458,16 @@
 	for(i = mx = tx = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
 		if(i < m->nmasters[m->curtag]) {
 			w = (m->ww - mx) / (MIN(n, m->nmasters[m->curtag]) - i);
-			resize(c, m->wx + mx, m->wy, w - (2*c->bw), mh - (2*c->bw), False);
-			mx += WIDTH(c);
+			if(n == 1)
+				resize(c, m->wx - c->bw, m->wy, m->ww, m->wh, False);
+			else
+				resize(c, m->wx + mx - c->bw, m->wy, w, mh - c->bw, False);
+			mx += WIDTH(c) - c->bw;
 		}
 		else {
 			w = (m->ww - tx) / (n - i);
-			resize(c, m->wx + tx, m->wy + mh, w - (2*c->bw), m->wh - mh - (2*c->bw), False);
-			tx += WIDTH(c);
+			resize(c, m->wx + tx - c->bw, m->wy + mh, w, m->wh - mh - c->bw, False);
+			tx += WIDTH(c) - c->bw;
 		}
 }
 
@@ -526,7 +530,7 @@
 	/* buddy list */
 	ww = (bl ? m->mfacts[m->curtag] * m->ww : m->ww);
 	if(bl) {
-		resize(bl, m->wx + ww, m->wy, m->ww - ww - 2 * bl->bw, m->wh - 2 * bl->bw, False);
+		resize(bl, m->wx - bl->bw + ww, m->wy, m->ww - ww, m->wh - bl->bw, False);
 		if(--n == 0)
 			return;
 	}
@@ -551,9 +555,12 @@
 		if(i/rows + 1 > cols - n%cols)
 			rows = n/cols + 1;
 		ch = rows ? m->wh / rows : m->wh;
-		cx = m->wx + cn*cw;
+		cx = m->wx + cn*cw - c->bw;
 		cy = m->wy + rn*ch;
-		resize(c, cx, cy, cw - 2 * c->bw, ch - 2 * c->bw, False);
+		if(n == 1 && !bl)
+			resize(c, cx, cy, cw, ch, False);
+		else
+			resize(c, cx, cy, cw - c->bw, ch - c->bw, False);
 		rn++;
 		if(rn >= rows) {
 			rn = 0;
@@ -990,6 +997,8 @@
 		attachstack(c);
 		grabbuttons(c, True);
 		XSetWindowBorder(dpy, c->win, dc.colors[1][ColBorder]);
+		XRaiseWindow(dpy, c->win);
+		raisefloating();
 		setfocus(c);
 	}
 	else
@@ -1334,7 +1343,7 @@
 	if(n > 0) /* override layout symbol */
 		snprintf(m->ltsymbol, sizeof m->ltsymbol, "[%d]", n);
 	for(c = nexttiled(m->clients); c; c = nexttiled(c->next))
-		resize(c, m->wx, m->wy, m->ww - 2 * c->bw, m->wh - 2 * c->bw, False);
+		resize(c, m->wx - c->bw, m->wy, m->ww, m->wh, False);
 }
 
 void
@@ -1503,6 +1512,20 @@
 	running = False;
 }
 
+void
+raisefloating(void) {
+	Client *c;
+	XEvent ev;
+
+	if(selmon->lt[selmon->sellt]->arrange) {
+		for(c = selmon->clients; c; c = c->next)
+			if(c->isfloating)
+				XRaiseWindow(dpy, c->win);
+	}
+	XSync(dpy, False);
+	while(XCheckMaskEvent(dpy, EnterWindowMask, &ev));
+}
+
 Monitor *
 recttomon(int x, int y, int w, int h) {
 	Monitor *m, *r = selmon;
@@ -1907,17 +1930,21 @@
 		mw = m->nmasters[m->curtag] ? m->ww * m->mfacts[m->curtag] : 0;
 	else
 		mw = m->ww;
-	for(i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
+	for(i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++) {
 		if(i < m->nmasters[m->curtag]) {
 			h = (m->wh - my) / (MIN(n, m->nmasters[m->curtag]) - i);
-			resize(c, m->wx, m->wy + my, mw - (2*c->bw), h - (2*c->bw), False);
-			my += HEIGHT(c);
+			if(n == 1)
+				resize(c, m->wx - c->bw, m->wy, m->ww, m->wh, False);
+			else
+				resize(c, m->wx - c->bw, m->wy + my, mw - c->bw, h - c->bw, False);
+			my += HEIGHT(c) - c->bw;
 		}
 		else {
 			h = (m->wh - ty) / (n - i);
-			resize(c, m->wx + mw, m->wy + ty, m->ww - mw - (2*c->bw), h - (2*c->bw), False);
-			ty += HEIGHT(c);
+			resize(c, m->wx + mw - c->bw, m->wy + ty, m->ww - mw, h - c->bw, False);
+			ty += HEIGHT(c) - c->bw;
 		}
+	}
 }
 
 void
