--- dwm.c.orig	2012-11-22 23:21:30.536217615 +0100
+++ b/dwm-6.0/dwm.c	2012-11-22 23:23:10.746234672 +0100
@@ -1325,9 +1325,14 @@
 		if(ISVISIBLE(c))
 			n++;
 	if(n > 0) /* override layout symbol */
-		snprintf(m->ltsymbol, sizeof m->ltsymbol, "[%d]", n);
-	for(c = nexttiled(m->clients); c; c = nexttiled(c->next))
-		resize(c, m->wx, m->wy, m->ww - 2 * c->bw, m->wh - 2 * c->bw, False);
+		snprintf(m->ltsymbol, sizeof m->ltsymbol, "   %d", n);
+	for(c = nexttiled(m->clients); c; c = nexttiled(c->next)) {
+		if(c->bw) {
+			c->oldbw = c->bw;
+			c->bw = 0;
+		}
+		resize(c, m->wx, m->wy, m->ww - (2 * c->bw), m->wh - (2 * c->bw), False);
+	}
 }
 
 void
@@ -1911,7 +1916,11 @@
 		mw = m->nmasters[m->curtag] ? m->ww * m->mfacts[m->curtag] : 0;
 	else
 		mw = m->ww;
-	for(i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
+	for(i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++) {
+		if(!c->bw && c->oldbw) {
+			c->bw = c->oldbw;
+			c->oldbw = 0;
+		}
 		if(i < m->nmasters[m->curtag]) {
 			h = (m->wh - my) / (MIN(n, m->nmasters[m->curtag]) - i);
 			resize(c, m->wx, m->wy + my, mw - (2*c->bw), h - (2*c->bw), False);
@@ -1922,6 +1931,7 @@
 			resize(c, m->wx + mw, m->wy + ty, m->ww - mw - (2*c->bw), h - (2*c->bw), False);
 			ty += HEIGHT(c) + gap;
 		}
+	}
 }
 
 void
@@ -1948,6 +1958,10 @@
 	if(!selmon->sel)
 		return;
 	selmon->sel->isfloating = !selmon->sel->isfloating || selmon->sel->isfixed;
+	if(selmon->sel->isfloating) {
+		selmon->sel->bw = selmon->sel->oldbw;
+		selmon->sel->oldbw = 0;
+	}
 	if(selmon->sel->isfloating)
 		resize(selmon->sel, selmon->sel->x, selmon->sel->y,
 		       selmon->sel->w, selmon->sel->h, False);
@@ -2494,7 +2508,11 @@
 		mh = m->nmasters[m->curtag] ? m->wh * m->mfacts[m->curtag] : 0;
 	else
 		mh = m->wh;
-	for(i = mx = tx = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
+	for(i = mx = tx = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++) {
+		if(!c->bw && c->oldbw) {
+			c->bw = c->oldbw;
+			c->oldbw = 0;
+		}
 		if(i < m->nmasters[m->curtag]) {
 			w = (m->ww - mx) / (MIN(n, m->nmasters[m->curtag]) - i);
 			resize(c, m->wx + mx, m->wy, w - (2*c->bw), mh - (2*c->bw), False);
@@ -2505,6 +2523,7 @@
 			resize(c, m->wx + tx, m->wy + mh, w - (2*c->bw), m->wh - mh - (2*c->bw), False);
 			tx += WIDTH(c) + gap;
 		}
+	}
 }
 
 int
