--- dwm.c.orig	2012-08-23 12:43:06.497197001 +0200
+++ b/dwm-6.0/dwm.c	2012-08-23 12:45:57.187194420 +0200
@@ -850,20 +850,16 @@
 	drawtext(m->ltsymbol, dc.colors[0], False);
 	dc.x += dc.w;
 	x = dc.x;
-	if(m == selmon) { /* status is only drawn on selected monitor */
-		dc.w = textnw(stext, strlen(stext)); // no padding
-		dc.x = m->ww - dc.w;
+	dc.w = textnw(stext, strlen(stext)); // no padding
+	dc.x = m->ww - dc.w;
 		if(showsystray && m == selmon) {
 			dc.x -= getsystraywidth();
 		}
-		if(dc.x < x) {
-			dc.x = x;
-			dc.w = m->ww - x;
-		}
-		drawcoloredtext(stext);
+	if(dc.x < x) {
+		dc.x = x;
+		dc.w = m->ww - x;
 	}
-	else
-		dc.x = m->ww;
+	drawcoloredtext(stext);
 	if((dc.w = dc.x - x) > bh) {
 		dc.x = x;
 		if(m->sel) {
@@ -2218,9 +2214,12 @@
 
 void
 updatestatus(void) {
+	Monitor* m;
+
 	if(!gettextprop(root, XA_WM_NAME, stext, sizeof(stext)))
 		strcpy(stext, "dwm-"VERSION);
-	drawbar(selmon);
+	for(m = mons; m; m = m->next)
+		drawbar(m);
 }
 
 void
