--- dwm.c	2013-03-08 20:28:06.666619655 +0100
+++ dwm.c	2013-03-08 20:25:35.813288858 +0100
@@ -2123,7 +2123,9 @@
 
 void
 view(const Arg *arg) {
+	int di;
 	unsigned int i;
+	Window win, dummy;
 
 	if((arg->ui & TAGMASK) == selmon->tagset[selmon->seltags])
 		return;
@@ -2145,6 +2147,10 @@
 	selmon->lt[selmon->sellt]= selmon->lts[selmon->curtag];
 	focus(NULL);
 	arrange(selmon);
+
+	XQueryPointer(dpy, root, &dummy, &win, &di, &di, &di, &di, &i);
+	focus(wintoclient(win));
+	XFlush(dpy);
 }
 
 Client *
