--- dwm.c.orig	2013-03-28 14:24:05.916058185 +0100
+++ dwm.c	2013-03-28 15:26:57.679328067 +0100
@@ -69,6 +69,7 @@
 	int i;
 	unsigned int ui;
 	float f;
+	float sf;
 	const void *v;
 } Arg;
 
@@ -144,6 +145,7 @@
 	int prevtag;
 	const Layout **lts;
 	double *mfacts;
+	double *smfacts;
 	int *nmasters;
 };
 
@@ -152,6 +154,7 @@
 	const Layout *layout;
 	float mfact;
 	int nmaster;
+	float smfact;
 } Tag;
 
 typedef struct {
@@ -236,6 +239,7 @@
 static void setfullscreen(Client *c, Bool fullscreen);
 static void setlayout(const Arg *arg);
 static void setmfact(const Arg *arg);
+static void setsmfact(const Arg *arg);
 static void setup(void);
 static int shifttag(int dist);
 static void showhide(Client *c);
@@ -452,7 +456,7 @@
 
 void
 bstack(Monitor *m) {
-	unsigned int i, n, w, mh, mx, tx;
+	unsigned int i, n, w, smw, mh, mx, tx;
 	Client *c;
 
 	for(n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), n++);
@@ -473,9 +477,19 @@
 			mx += WIDTH(c) - c->bw;
 		}
 		else {
-			w = (m->ww - tx) / (n - i);
-			resize(c, m->wx + tx - c->bw, m->wy + mh, w, m->wh - mh - c->bw, False);
-			tx += WIDTH(c) - c->bw;
+			smw = m->mh * m->smfacts[m->curtag];
+			if(!(nexttiled(c->next)))
+				w = (m->ww - tx) / (n - i);
+			else
+				w = (m->ww - smw - tx) / (n - i);
+			if(h > minwsz)
+				resize(c, m->wx + tx - c->bw, m->wy + mh, w, m->wh - mh - c->bw, False);
+			else
+				c->isfloating = True;
+			if(!(nexttiled(c->next)))
+				tx += WIDTH(c) - c->bw + smw;
+			else
+				tx += WIDTH(c) - c->bw;
 		}
 }
 
@@ -627,6 +641,7 @@
 	XUnmapWindow(dpy, mon->barwin);
 	XDestroyWindow(dpy, mon->barwin);
 	free(mon->mfacts);
+	free(mon->smfacts);
 	free(mon->nmasters);
 	free(mon->lts);
 	free(mon);
@@ -768,12 +783,15 @@
 		die("fatal: could not malloc() %u bytes\n", sizeof(Monitor));
 	if(!(m->mfacts = calloc(numtags, sizeof(double))))
 		die("fatal: could not malloc() %u bytes\n", sizeof(double) * numtags);
+	if(!(m->smfacts = calloc(numtags, sizeof(double))))
+		die("fatal: could not malloc() %u bytes\n", sizeof(double) * numtags);
 	if(!(m->nmasters = calloc(numtags, sizeof(int))))
 		die("fatal: could not malloc() %u bytes\n", sizeof(int) * numtags);
 	if(!(m->lts = calloc(numtags, sizeof(Layout *))))
 		die("fatal: could not malloc() %u bytes\n", sizeof(Layout *) * numtags);
 	m->tagset[0] = m->tagset[1] = 1;
 	m->mfacts[0] = mfact;
+	m->smfacts[0] = smfact;
 	m->nmasters[0] = nmaster;
 	m->lts[0] = &layouts[0];
 	m->showbar = showbar;
@@ -781,6 +799,7 @@
 	m->curtag = m->prevtag = 1;
 	for(i = 1; i < numtags; i++) {
 		m->mfacts[i] = tags[i - 1].mfact < 0 ? mfact : tags[i - 1].mfact;
+		m->smfacts[i] = tags[i - 1].smfact < 0 ? smfact : tags[i -1].smfact;
 		m->nmasters[i] = tags[i - 1].nmaster < 0 ? nmaster : tags[i - 1].nmaster;
 		m->lts[i] = tags[i - 1].layout;
 	}
@@ -1766,6 +1785,19 @@
 }
 
 void
+setsmfact(const Arg *arg) {
+	float sf;
+
+	if(!arg || !selmon->lt[selmon->sellt]->arrange)
+		return;
+	sf = arg->sf < 1.0 ? arg->sf + selmon->smfacts[selmon->curtag] : arg->sf - 1.0;
+	if(sf < 0 || sf > 0.9)
+		return;
+	selmon->smfacts[selmon->curtag] = sf;
+	arrange(selmon);
+}
+
+void
 setup(void) {
 	XSetWindowAttributes wa;
 
@@ -1906,7 +1938,7 @@
 
 void
 tile(Monitor *m) {
-	unsigned int i, n, h, mw, my, ty;
+	unsigned int i, n, h, smh, mw, my, ty;
 	Client *c;
 
 	for(n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), n++);
@@ -1927,9 +1959,19 @@
 			my += HEIGHT(c) - c->bw;
 		}
 		else {
-			h = (m->wh - ty) / (n - i);
-			resize(c, m->wx + mw - c->bw, m->wy + ty, m->ww - mw, h - c->bw, False);
-			ty += HEIGHT(c) - c->bw;
+			smh = m->mh * m->smfacts[m->curtag];
+			if(!(nexttiled(c->next)))
+				h = (m->wh - ty) / (n - i);
+			else
+				h = (m->wh - smh - ty) / (n - i);
+			if(h > minwsz)
+				resize(c, m->wx + mw - c->bw, m->wy + ty, m->ww - mw, h - c->bw, False);
+			else
+				c->isfloating = True;
+			if(!(nexttiled(c->next)))
+				ty += HEIGHT(c) - c->bw + smh;
+			else
+				ty += HEIGHT(c) - c->bw;
 		}
 }
 
