--- dwm.c.orig	2013-03-27 16:33:58.646639392 +0100
+++ dwm.c	2013-03-27 16:35:28.446637881 +0100
@@ -126,6 +126,7 @@
 
 struct Monitor {
 	char ltsymbol[16];
+	int status;
 	float mfact;
 	int nmaster;
 	int num;
@@ -650,6 +651,7 @@
 	if(!(m = (Monitor *)calloc(1, sizeof(Monitor))))
 		die("fatal: could not malloc() %u bytes\n", sizeof(Monitor));
 	m->tagset[0] = m->tagset[1] = 1;
+	m->status = 0;
 	m->mfact = mfact;
 	m->nmaster = nmaster;
 	m->showbar = showbar;
@@ -740,7 +742,7 @@
 	drawtext(m->ltsymbol, dc.norm, False);
 	dc.x += dc.w;
 	x = dc.x;
-	if(m == selmon) { /* status is only drawn on selected monitor */
+	if(m->status) { /* status is only drawn on status monitor */
 		dc.w = TEXTW(stext);
 		dc.x = m->ww - dc.w;
 		if(dc.x < x) {
@@ -1888,6 +1890,7 @@
 					m->my = m->wy = unique[i].y_org;
 					m->mw = m->ww = unique[i].width;
 					m->mh = m->wh = unique[i].height;
+					m->status = (i == statusmon ? 1 : 0);
 					updatebarpos(m);
 				}
 		}
@@ -1916,6 +1919,7 @@
 	{
 		if(!mons)
 			mons = createmon();
+		mons->status = 1;
 		if(mons->mw != sw || mons->mh != sh) {
 			dirty = True;
 			mons->mw = mons->ww = sw;
@@ -2005,9 +2009,12 @@
 
 void
 updatestatus(void) {
+	Monitor *m;
+
 	if(!gettextprop(root, XA_WM_NAME, stext, sizeof(stext)))
 		strcpy(stext, "dwm-"VERSION);
-	drawbar(selmon);
+	for(m = mons; m; m = m->next)
+		drawbar(m);
 }
 
 void
