--- dwm.c.orig	2013-04-06 14:11:27.377910274 +0200
+++ dwm.c	2013-04-06 14:12:14.204576153 +0200
@@ -1337,7 +1337,7 @@ monocle(Monitor *m) {
 		if(ISVISIBLE(c))
 			n++;
 	if(n > 0) /* override layout symbol */
-		snprintf(m->ltsymbol, sizeof m->ltsymbol, "[%d]", n);
+		snprintf(m->ltsymbol, sizeof m->ltsymbol, " %d", n);
 	for(c = nexttiled(m->clients); c; c = nexttiled(c->next))
 		resize(c, m->wx - c->bw, m->wy, m->ww, m->wh, False);
 }
